--- a/drivers/gpu/drm/amd/amdgpu/amdgpu.h	2026-06-09 10:32:51.000000000 +0000
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu.h	2026-06-20 18:32:23.484183052 +0000
@@ -157,6 +157,7 @@
 /*
  * Modules parameters.
  */
+extern int amdgpu_bc250;
 extern int amdgpu_modeset;
 extern unsigned int amdgpu_vram_limit;
 extern int amdgpu_vis_vram_limit;
--- a/drivers/gpu/drm/amd/amdgpu/amdgpu_drv.c	2026-06-09 10:32:51.000000000 +0000
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu_drv.c	2026-06-20 18:31:48.981086644 +0000
@@ -272,6 +272,10 @@
 	.period = 0x0, /* default to 0x0 (timeout disable) */
 };
 
+int amdgpu_bc250 = 0;
+MODULE_PARM_DESC(bc250, "Enable BC250 Patches");
+module_param_named(bc250, amdgpu_bc250, int, 0644);
+
 /**
  * DOC: vramlimit (int)
  * Restrict the total amount of VRAM in MiB for testing.  The default is 0 (Use full VRAM).
--- a/drivers/gpu/drm/amd/amdgpu/gmc_v10_0.c	2026-06-09 10:32:51.000000000 +0000
+++ b/drivers/gpu/drm/amd/amdgpu/gmc_v10_0.c	2026-06-20 18:33:27.145245964 +0000
@@ -766,6 +766,10 @@
 	} else if (amdgpu_emu_mode == 1) {
 		adev->gmc.vram_type = AMDGPU_VRAM_TYPE_GDDR6;
 		adev->gmc.vram_width = 1 * 128; /* numchan * chansize */
+	} else if (amdgpu_bc250 == 1) {
+		adev->gmc.is_app_apu = false;
+		adev->gmc.vram_type = AMDGPU_VRAM_TYPE_GDDR6;
+		adev->gmc.vram_width = 256;
 	} else {
 		r = amdgpu_atomfirmware_get_vram_info(adev,
 				&vram_width, &vram_type, &vram_vendor);
@@ -970,6 +974,10 @@
 	int r;
 
 	adev->gmc.flush_pasid_uses_kiq = !amdgpu_emu_mode;
+	if (amdgpu_bc250 == 1) {
+	        adev->gmc.flush_pasid_uses_kiq = false;
+	}
+
 
 	/* The sequence of these two function calls matters.*/
 	gmc_v10_0_init_golden_registers(adev);
--- a/drivers/gpu/drm/amd/amdkfd/kfd_crat.c	2026-06-09 10:32:51.000000000 +0000
+++ b/drivers/gpu/drm/amd/amdkfd/kfd_crat.c	2026-06-20 18:15:45.467336886 +0000
@@ -1125,6 +1125,11 @@
 							mem->length_low;
 			width = mem->width;
 
+			if (amdgpu_bc250 == 1) {
+				width = 256;
+				heap_type = HSA_MEM_HEAP_TYPE_FB_PUBLIC;
+			}
+
 			/* Multiple banks of the same type are aggregated into
 			 * one. User mode doesn't care about multiple physical
 			 * memory segments. It's managed as a single virtual
